galley/pkg/runtime/state.go

```diff
@@ -147,13 +152,92 @@ func (s *State) buildSnapshot() snapshot.Snapshot {
 	// Build entities that are derived from existing ones.
 	s.buildProjections(b)
 
+	if ret, _ := strconv.ParseBool(os.Getenv("ISTIO_CONVERT_AUTH_POLICY")); ret {
+		// Update AuthenticationPolicy to use labels in its target selector.
+		s.updateAuthenticationPolicy(b)
+	}
+
 	return b.Build()
 }
 
 func (s *State) buildProjections(b *snapshot.InMemoryBuilder) {
 	s.buildIngressProjectionResources(b)
 }
 
+func (s *State) createAuthConverter() *conversions.AuthConverter {
+	serviceState := s.entries[metadata.Service.Collection]
+	if serviceState == nil {
+		return nil
+	}
+
+	authConverter := conversions.AuthConverter{}
+	for name, entry := range serviceState.entries {
+		selector := make(resource.Labels)
+		data := entry.Metadata.Annotations["_service_spec_selector_"]
+		if err := json.Unmarshal([]byte(data), &selector); err != nil {
+			scope.Errorf("failed to unmarshal service %s: %v", name, err)
+			continue
+		}
+		authConverter.AddService(name, selector)
+	}
+
+	scope.Debugf("created auth converter: %+v", authConverter)
+	return &authConverter
+}
+
+func (s *State) updateAuthenticationPolicy(b *snapshot.InMemoryBuilder) {
+	authConverter := s.createAuthConverter()
+	if authConverter == nil {
+		return
+	}
+
+	state := s.entries[metadata.Policy.Collection]
+	if state == nil {
+		return
+	}
+
+	for name, entry := range state.entries {
+		policy, err := conversions.ToAuthenticationPolicy(entry)
+		if err != nil {
+			scope.Errorf("error updating authentication policy: %v", err)
+			return
+		}
+
+		namespace, _ := name.InterpretAsNamespaceAndName()
+		for _, target := range policy.Targets {
+			selectors := authConverter.GetSelectors(target.Name, namespace, false)
+			if len(selectors) != 1 {
+				scope.Debugf("no services for authentication policy %s/%s", namespace, target.Name)
+				continue
+			}
+			scope.Debugf("matched authentication policy %s with selectors: %v", target.Name, selectors)
+			target.Labels = selectors[0]
+
+			// Convert ports to labels. e.g. ports [{name: "http"} , {number: 80}] would be converted to
+			// labels { "__port_name_0": "http", "__port_number_1": "80" }.
```

Comment:
Do we need to convert ports to labels? This seems unnatural. Can we use the existing "ports" field in "target"?

Reply:
@diemtvu 
I remember you mentioned that we will need ports in the labels, do we still need this now?